### ISSUE TITLE
Failing Kokkos build on Debian

### ISSUE THREAD

**Issue by crtrott** [kokkos/kokkos#402]
This is from an email exchange on the kokkos list:
```
Hi,

  unfortunately the machine with gcc 4.85 belongs to a colleague who just 
left the office. I quickly used another computer I have access to. The 
compiler is gcc 4.9.2:

g++ (Debian 4.9.2-10) 4.9.2

I pulled the most recent version of Kokkos available and configured with 
OpenMP as the only device as before. KokkosCore_config.h is still not 
located correctly and the beginning of the error is the same as before:

Scanning dependencies of target kokkos_test
[100%] Building CXX object 
CMakeFiles/kokkos_test.dir/src/kokkos_openmp.cpp.o
In file included from 
/home/mvymaza1/local/kokkos-openmp/include/Kokkos_ExecPolicy.hpp:48:0,
                  from 
/home/mvymaza1/local/kokkos-openmp/include/Kokkos_MemoryPool.hpp:48,
                  from 
/home/mvymaza1/local/kokkos-openmp/include/Kokkos_Core.hpp:68,
                  from 
/home/mvymaza1/kokkos_openmp/src/kokkos_openmp.cpp:1:
/home/mvymaza1/local/kokkos-openmp/include/impl/Kokkos_Traits.hpp:380:23: 
error: ISO C++ forbids declaration of ‘type name’ with no type 
[-fpermissive]
  struct is_label<const std::string> : public true_type {};
                        ^
/home/mvymaza1/local/kokkos-openmp/include/impl/Kokkos_Traits.hpp:380:34: 
error: template argument 1 is invalid
  struct is_label<const std::string> : public true_type {};
                                   ^
/home/mvymaza1/local/kokkos-openmp/include/impl/Kokkos_Traits.hpp:383:17: 
error: ‘string’ is not a member of ‘std’
  struct is_label<std::string> : public true_type {};
                  ^
/home/mvymaza1/local/kokkos-openmp/include/impl/Kokkos_Traits.hpp:383:17: 
error: ‘string’ is not a member of ‘std’
/home/mvymaza1/local/kokkos-openmp/include/impl/Kokkos_Traits.hpp:383:28: 
error: template argument 1 is invalid
  struct is_label<std::string> : public true_type {};
                             ^
In file included from 
/home/mvymaza1/local/kokkos-openmp/include/Kokkos_MemoryPool.hpp:48:0,
                  from 
/home/mvymaza1/local/kokkos-openmp/include/Kokkos_Core.hpp:68,
                  from 
/home/mvymaza1/kokkos_openmp/src/kokkos_openmp.cpp:1:
/home/mvymaza1/local/kokkos-openmp/include/Kokkos_ExecPolicy.hpp: In 
member function ‘void 
Kokkos::RangePolicy<Properties>::set_auto_chunk_size()’:
/home/mvymaza1/local/kokkos-openmp/include/Kokkos_ExecPolicy.hpp:312:12: 
error: ‘abort’ is not a member of ‘Kokkos’
             Kokkos::abort("RangePolicy blocking granularity must be 
power of two" );
             ^
/home/mvymaza1/local/kokkos-openmp/include/Kokkos_ExecPolicy.hpp:312:12: 
note: suggested alternatives:
In file included from /usr/include/c++/4.9/cstdlib:72:0,
                  from /usr/include/c++/4.9/ext/string_conversions.h:41,
                  from /usr/include/c++/4.9/bits/basic_string.h:2850,
                  from /usr/include/c++/4.9/string:52,
                  from /usr/include/c++/4.9/bits/locale_classes.h:40,
                  from /usr/include/c++/4.9/bits/ios_base.h:41,
                  from /usr/include/c++/4.9/ios:42,
                  from /usr/include/c++/4.9/ostream:38,
                  from /usr/include/c++/4.9/iostream:39,
                  from 
/home/mvymaza1/local/kokkos-openmp/include/Kokkos_ExecPolicy.hpp:51,
                  from 
/home/mvymaza1/local/kokkos-openmp/include/Kokkos_MemoryPool.hpp:48,
                  from 
/home/mvymaza1/local/kokkos-openmp/include/Kokkos_Core.hpp:68,
                  from 
/home/mvymaza1/kokkos_openmp/src/kokkos_openmp.cpp:1:
/usr/include/stdlib.h:515:13: note:   ‘abort’
  extern void abort (void) __THROW __attribute__ ((__noreturn__));
              ^
/usr/include/stdlib.h:515:13: note:   ‘abort’
In file included from 
/home/mvymaza1/local/kokkos-openmp/include/Kokkos_Core.hpp:68:0,
                  from 
/home/mvymaza1/kokkos_openmp/src/kokkos_openmp.cpp:1:
/home/mvymaza1/local/kokkos-openmp/include/Kokkos_MemoryPool.hpp: In 
constructor ‘Kokkos::Experimental::MempoolImpl::initialize_array<T, 
ExecutionSpace>::initialize_array(T*, size_t, T)’:
/home/mvymaza1/local/kokkos-openmp/include/Kokkos_MemoryPool.hpp:96:5: 
error: ‘parallel_for’ is not a member of ‘Kokkos’
      Kokkos::parallel_for( size, *this );

It does not seem to be a linker error, but rather a missing header 
somewhere (std::string is not recognized). I can send you the complete 
log file if you think that would help (I'm not sure I can attach 
anything to mailing list posts, so I'd rather send that to your personal 
address or upload somewhere).

Thank you very much.

Best regards,

  Martin
```

**Comment by crtrott** [comment]
OK If you have a moment lets try and figure this out. As I said I don't see the error on hour red-hat machines, but a common cause for this type of thing is that system header files include each other on some platforms but not on others. So first thing to try:

1) add an "#include <string>" in core/src/impl/Kokkos_Traits.hpp

**Comment by martinv** [comment]
I managed to eliminate almost all errors by adding headers:

```
diff --git a/core/src/Kokkos_ExecPolicy.hpp b/core/src/Kokkos_ExecPolicy.hpp
index d8d6199..ab97761 100644
--- a/core/src/Kokkos_ExecPolicy.hpp
+++ b/core/src/Kokkos_ExecPolicy.hpp
@@ -48,6 +48,7 @@
 #include <impl/Kokkos_Traits.hpp>
 #include <impl/Kokkos_StaticAssert.hpp>
 #include <impl/Kokkos_Tags.hpp>
+#include <impl/Kokkos_Error.hpp>
 #include <iostream>
 //----------------------------------------------------------------------------
 
diff --git a/core/src/Kokkos_MemoryPool.hpp b/core/src/Kokkos_MemoryPool.hpp
index f0e4c36..b474ac2 100644
--- a/core/src/Kokkos_MemoryPool.hpp
+++ b/core/src/Kokkos_MemoryPool.hpp
@@ -47,6 +47,7 @@
 #include <Kokkos_Core_fwd.hpp>
 #include <Kokkos_ExecPolicy.hpp>
 #include <Kokkos_Atomic.hpp>
+#include <Kokkos_Parallel.hpp>
 #include <impl/Kokkos_BitOps.hpp>
 #include <impl/Kokkos_Error.hpp>
 #include <impl/KokkosExp_SharedAlloc.hpp>
diff --git a/core/src/impl/Kokkos_Traits.hpp b/core/src/impl/Kokkos_Traits.hpp
index dc8231b..8664330 100644
--- a/core/src/impl/Kokkos_Traits.hpp
+++ b/core/src/impl/Kokkos_Traits.hpp
@@ -48,6 +48,7 @@
 #include <stdint.h>
 #include <Kokkos_Macros.hpp>
 #include <type_traits>
+#include <string>
 
 namespace Kokkos {
 namespace Impl {
```

and the error is now much shorter:

```
/home/mvymaza1/local/kokkos-openmp/include/Kokkos_Core.hpp:121:65: error: ‘memory_space’ in ‘Kokkos::DefaultExecutionSpace {aka class Kokkos::OpenMP}’ does not name a type
 template< class Space = typename Kokkos::DefaultExecutionSpace::memory_space >
                                                                 ^
/home/mvymaza1/local/kokkos-openmp/include/Kokkos_Core.hpp:131:65: error: ‘memory_space’ in ‘Kokkos::DefaultExecutionSpace {aka class Kokkos::OpenMP}’ does not name a type
 template< class Space = typename Kokkos::DefaultExecutionSpace::memory_space >
                                                                 ^
/home/mvymaza1/local/kokkos-openmp/include/Kokkos_Core.hpp:140:65: error: ‘memory_space’ in ‘Kokkos::DefaultExecutionSpace {aka class Kokkos::OpenMP}’ does not name a type
 template< class Space = typename Kokkos::DefaultExecutionSpace::memory_space >
                                                                 ^
/home/mvymaza1/local/kokkos-openmp/include/Kokkos_Core.hpp:149:65: error: ‘memory_space’ in ‘Kokkos::DefaultExecutionSpace {aka class Kokkos::OpenMP}’ does not name a type
 template< class Space = typename Kokkos::DefaultExecutionSpace::memory_space >
                                                                 ^
CMakeFiles/kokkos_test.dir/build.make:54: recipe for target 'CMakeFiles/kokkos_test.dir/src/kokkos_openmp.cpp.o' failed
make[2]: *** [CMakeFiles/kokkos_test.dir/src/kokkos_openmp.cpp.o] Error 1
CMakeFiles/Makefile2:60: recipe for target 'CMakeFiles/kokkos_test.dir/all' failed
make[1]: *** [CMakeFiles/kokkos_test.dir/all] Error 2
```

Apparently the compiler does not see Kokkos::OpenMP in Kokkos_Core.hpp, but I don't understand what's wrong, because Kokkos_Core.hpp includes Kokkos_Core_fwd.hpp, which pulls Kokkos_Macros.hpp. Kokkos_Macros.hpp include KokkosCore_config.h whose contents is as follows:
```
/* ---------------------------------------------
Makefile constructed configuration:
Fri 19 Aug 20:13:26 BST 2016
----------------------------------------------*/
/* Execution Spaces */
#define KOKKOS_HAVE_OPENMP 1
/* General Settings */
#define KOKKOS_HAVE_CXX11 1
/* Optimization Settings */
/* Cuda Settings */
```

This means that in Kokkos_Core.hpp, the directive
```
#if defined( KOKKOS_HAVE_OPENMP )
#include <Kokkos_OpenMP.hpp>
#endif
```
includes the OpenMP header. Do I need any other headers for this to compile?

Best regards,

  Martin

**Comment by crtrott** [comment]
can you check how the develop branch behaves for you. I just saw that the first two includes are already in develop. Sorry that this is such a hastle. One of us will try to replicate the issue on his private machine (with Ubuntu) at home later. So far we were unable to replicate it on any of our machines here at Sandia.

**Comment by martinv** [comment]
I just tested the develop branch. I don't need to apply any of the fixes above, but I still have this one last compilation error (error: ‘memory_space’ in ‘Kokkos::DefaultExecutionSpace {aka class Kokkos::OpenMP}’ does not name a type).

I'll eventually try to play with it and see if I can fix it somehow.

Just out of curiosity, what (Linux) systems do you develop and test on? Do you use only RedHat or do you consider other distributions as well?

**Comment by crtrott** [comment]
We are currently running red-hat only across a number of platforms (Power8, Xeon, Xeon Phi), its a bit tricky for us to test on other systems due to security approval reasons. 

Can you test this little code: 
```C++
#include<cstdio>
int main () {
#ifdef _OPENMP
  printf("Have OpenMP\n");
#else
  printf("No OpenMP\n");
#endif
}
```
with both g++ -fopenmp and just g++.

**Comment by martinv** [comment]
HI Christian,

 the above prints "No OpenMP" unless I compile the code with -fopenmp (in which case I see "Have OpenMP"). This flag is apparently automatically added when Kokkos is built with OpenMP support, so I tried to force the same flag when building my executable before linking to Kokkos (develop branch) and adding -fopenmp to the list of CXX flags indeed seems to solve the problem. Thank you very much for your help.

I will make more tests on other machines on Monday, but I suppose that everything will be fine now.

Best regards,

 Martin

**Comment by crtrott** [comment]
Ok so the link issue goes away with adding the flags? Great. Btw, when linking against a standalone build of Kokkos you should still include the Makefile.kokkos in your Makefile (if you use Makefiles) and then use the variables it provides to append your flags. This is documented in the Kokkos Programming guide in the kokkos/doc directory. If you are using CMake or so for your bigger app its unfortunately more complicated.

**Comment by martinv** [comment]
Yes, I used CMake both for my little test application and the big one. Speaking about that, since the big application (Nektar++) depends on a number of libraries and some of them are shared, we have to make sure that Kokkos is compiled with -fPIC, otherwise the linker complains.

Is there any plan for Kokkos to directly support CMake? I suppose that that might be just a bit too hard to maintain and keep sufficiently robust on multiple platforms with libraries like CUDA.

**Comment by mhoemmen** [comment]
@martinv Kokkos does support CMake through Trilinos' build system.  It's possible to do a "sparse checkout" of Trilinos that only includes the Kokkos package.  If you're interested in that option, let us know.

**Comment by martinv** [comment]
We currently build Kokkos as a CMake external project. Nektar++ has its own linear solver and also an interface to Petsc and we plan to give Trilinos a go, so maybe in the future we'll consider this option. Thanks for letting me know.

**Comment by crtrott** [comment]
Args we still didn't put the string in include in. Adding bug label and self assign.

**Comment by crtrott** [comment]
Actually I was lying, the string include was in there, since late May. Are you really sure you were using a recent version of Kokkos? This is from git blame: Kokkos_Traits.hpp

```
^d0932b0 (Carter Edwards 2015-04-09 09:28:05 -0600  47) #include <stddef.h>
^d0932b0 (Carter Edwards 2015-04-09 09:28:05 -0600  48) #include <stdint.h>
^d0932b0 (Carter Edwards 2015-04-09 09:28:05 -0600  49) #include <Kokkos_Macros.hpp>
0b7d210c (Carter Edwards 2016-05-27 14:45:46 -0600  50) #include <string>
e1b228b2 (crtrott        2016-01-14 16:55:23 -0700  51) #include <type_traits>
```

as you can see the string include was added in May. Your diff did't have this.

**Comment by martinv** [comment]
Yes I also have the <string> header included in the master branch. This is strange so I checked what version of Kokkos I have on the Debian machine where I encountered some compilation difficulties. It turns out that the last commit I have there is 9afaa87fa37f4bc90913d7b31a3c168883ceda55 from Thursday, the 21st of July with comment 'Merge branch 'develop''. This commit does not seem to contain the <string> header, but since this was fixed later on, there's no need to worry about it I guess.
